### ISSUE TITLE
Arrays of Arrays are problematic in VCDM

### ISSUE THREAD

**Issue by melvincarvalho** [w3c/vc-data-model#1132]
I've been trying to model an array of arrays, which are used on Nostr for tags.  Working in the Nostr CG I am trying to translate the events (signed json blobs) to be compatible with VC

First of all its a challenge to find text and a `@context` to define these.  My understanding is that first you translate to RDF, then you sign.  This seems a heavy burden, especially with millions of events flying around (I currently have ~20 million locally) and space is an issue, as well as time to verify.  

An array of arrays In JSON
```json
"numbers": [[1, 2], [3, 4]]
```

I've been trying to figure out how this looks in VC.  My best guess so far is:

```
{
  "@context": {
    "@vocab": "http://example.org/",
    "rdf": "http://www.w3.org/1999/02/22-rdf-syntax-ns#"
  },
  "numbers": {
    "@id": "_:b0",
    "rdf:first": {
      "@id": "_:b1",
      "rdf:first": 1,
      "rdf:rest": {
        "@id": "_:b2",
        "rdf:first": 2,
        "rdf:rest": {
          "@id": "rdf:nil"
        }
      }
    },
    "rdf:rest": {
      "@id": "_:b3",
      "rdf:first": {
        "@id": "_:b4",
        "rdf:first": 3,
        "rdf:rest": {
          "@id": "_:b5",
          "rdf:first": 4,
          "rdf:rest": {
            "@id": "rdf:nil"
          }
        }
      },
      "rdf:rest": {
        "@id": "rdf:nil"
      }
    }
  }
}
```

This is quite a verbose expansion.

Aside from performance issues, and complexity, it was unclear to me how to define a bnode strategy, if indeed my data structure is correct.

Then how would you sign a credential which contains this JSON.

Interop is going to be important, and there is going to have to be a way to do this.  It seems quite a complex path right now.  Can there be alternatives?

**Comment by TallTed** [comment]
@melvincarvalho — Please return to your initial comment on this issue, #1132, and code fence the `@context` in the second paragraph (as, ``` `@context` ```), so that GitHub user is not dragged into our conversation without choosing to be so.

**Comment by melvincarvalho** [comment]
> @melvincarvalho — Please return to your initial comment on this issue, #1132, and code fence the `@context` in the second paragraph (as, `` `@context` ``), so that GitHub user is not dragged into our conversation without choosing to be so.

Thank you for bringing this to my attention. I have addressed your request this time. However, considering the user has been inactive since 2009, I believe it's a non-issue. Unless it's a specific group policy, I won't be concerned about it in the future.

**Comment by davidlehn** [comment]
I'm unclear on the issue here.  Are you trying to model *ordered* lists of *ordered* lists of values?  JSON-LD has native support for [ordered lists](https://www.w3.org/TR/json-ld/#lists) that are recursive.  Lists of geo coordinates was a use case when this was developed.  You don't have to expand into that RDF list structure yourself.  Tools will do that.

As a plain JSON-LD example ([playground](https://json-ld.org/playground/#startTab=tab-expanded&copyContext=true&json-ld=%7B%22%40context%22%3A%7B%22numbers%22%3A%7B%22%40id%22%3A%22ex%3Anumbers%22%2C%22%40container%22%3A%22%40list%22%7D%7D%2C%22numbers%22%3A%5B%5B1%2C2%5D%2C%5B3%2C4%5D%5D%7D)):
```json
{
  "@context": {
    "numbers": {
      "@id": "ex:numbers",
      "@container": "@list"
    }
  },
  "numbers": [[1,2],[3,4]]
}
```
That will [expand](https://www.w3.org/TR/json-ld11-api/#expansion) to a form with all the `@list` keywords.  If you [compact](https://www.w3.org/TR/json-ld11-api/#compaction) with the same context, it will stay as JSON arrays.  When serialized to N-Quads, or [canonicalize](https://www.w3.org/TR/rdf-canon/) for signing, it will have the more verbose RDF list form you have above.  Check the playground tabs above to see the outputs.

If your use case is unordered sets, there may be different issues to consider.  If you're mixing data structures... it can be tricky.

Current tooling can do all the transformations needed for signing and verifying.

Or am I misunderstanding and you are trying to go directly to a canonicalized form for signing as an optimziation?  That can be possible depending on your use case and data.  I've never tried this with more complex structures like lists.  But sometimes you can create a pre-canonicalized "template", fill in only changing values properly, then sign.

It might help to say a bit more about the use case.

**Comment by melvincarvalho** [comment]
@davidlehn I want to extend my sincere thanks for your insightful response. Your guidance has been incredibly helpful.

Previously, I had raised this topic on the [json-ld-syntax](https://github.com/w3c/json-ld-syntax/issues/414) list and the W3C Nostr CG [mailing list](https://lists.w3.org/Archives/Public/public-nostr/2023May/0002.html), and I've also had private discussions with Kingsley and Nathan. Your comprehensive answer has provided much-needed clarity on this matter.

One of the main challenges I've encountered is transitioning from the intuitive JSON `"numbers": [[1,2],[3,4]]` structure to the VCDM structure. Could you confirm if the VCDM structure is the canonical form that gets signed?

JSON

```json
"numbers": [[1,2],[3,4]]
```

To VCDM:

```turtle
_:b0 <ex:numbers> _:b1 .
_:b1 <http://www.w3.org/1999/02/22-rdf-syntax-ns#first> _:b2 .
_:b1 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> _:b4 .
_:b2 <http://www.w3.org/1999/02/22-rdf-syntax-ns#first> "1"^^<http://www.w3.org/2001/XMLSchema#integer> .
_:b2 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> _:b3 .
_:b3 <http://www.w3.org/1999/02/22-rdf-syntax-ns#first> "2"^^<http://www.w3.org/2001/XMLSchema#integer> .
_:b3 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> <http://www.w3.org/1999/02/22-rdf-syntax-ns#nil> .
_:b4 <http://www.w3.org/1999/02/22-rdf-syntax-ns#first> _:b5 .
_:b4 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> <http://www.w3.org/1999/02/22-rdf-syntax-ns#nil> .
_:b5 <http://www.w3.org/1999/02/22-rdf-syntax-ns#first> "3"^^<http://www.w3.org/2001/XMLSchema#integer> .
_:b5 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> _:b6 .
_:b6 <http://www.w3.org/1999/02/22-rdf-syntax-ns#first> "4"^^<http://www.w3.org/2001/XMLSchema#integer> .
_:b6 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> <http://www.w3.org/1999/02/22-rdf-syntax-ns#nil> .
```

Is this the canonical form that gets signed, or is it something else?

Visually depiction:

![image](https://github.com/w3c/vc-data-model/assets/65864/f800243a-87e4-44ec-a4a5-0dad2a61d110)

My primary use case involves using semantic interoperability to engage with standardized models in JSON(-LD) and/or VCDM, particularly in relation to the [nostr](https://github.com/nostr-protocol/nips/blob/master/01.md) protocol, where tags are portrayed as an array of arrays. I see potential for fruitful interaction and interoperability with Block's work, where there could be some intersection with semantic interoperabilty

An example use case here is signed badges, which are very popular in nostr:

![image](https://github.com/w3c/vc-data-model/assets/65864/cc85d494-b265-4477-a3ff-2d84fe3508ee)


The translation from JSON to VCDM still presents a significant challenge and there is an evident education gap. To further my understanding, I have several questions I'd like to explore:

- What is the canonical form of `"numbers": [[1,2],[3,4]]` in VCDM?
- What would the signing process look like?
- How can I better understand the bnode strategy (b0,b1,b2,b3,b4,b5,b6)?
- How does this strategy align to array indices (0,0 ... 1,1)?
- What are the performance implications?
- Is there a more effective way to tackle this?

This discussion aligns with the observation that ordered lists (arrays) are prevalent in JSON but rare in RDF. 

Once again, thank you for your invaluable assistance.

**Comment by OR13** [comment]
Afaik, arrays of arrays are not valid JSON-LD, since order is not preserved, this makes any credential containing arrays or arrays invalid... It's also a trivial way to attack the Canonicalization algorithm.... 

Bad news for the matrix.

**Comment by iherman** [comment]
The issue was discussed in a [meeting](https://www.w3.org/2017/vc/WG/Meetings/Minutes/2023-07-05-vcwg#section2-9) on 2023-07-05

- no resolutions were taken

<details><summary><i>View the transcript</i></summary>

#### 2.9. Arrays of Arrays are problematic in VCDM (issue vc-data-model#1132)

_See github issue [vc-data-model#1132](https://github.com/w3c/vc-data-model/issues/1132)._




> *Orie Steele:* Lets add an example that has a matrix in it :).

**Kristina Yasuda:** We can't find text in `@context` that defines how to do that.  

> *Orie Steele:* I object,.

**Kristina Yasuda:** Arrays of arrays are not valid JSON-LD.  
… I'm going to put pending close unless there are objections.  

> *Orie Steele:* it needs to be explained in the spec.

> *Michael Prorock:* +1 orie.

**Orie Steele:** Having answered this question many times, if our Core DM is JSON-LD, we're obliged to describe this detail.  
… If people try to do what works in JSON they will be really unhappy.  
… We should describe `@list` and `@type` and ordering.  
… That kind of guidance needs to be in our spec.  

> *Orie Steele:* We could recommend `@type`, `@json`, `@list`.

**Kristina Yasuda:** Could you take a stab at this, Orie?  

> *Orie Steele:* etc...

**Orie Steele:** I think it would be better for someone more experienced in JSON-LD to do that.  

**Manu Sporny:** You can do arrays of arrays and arrays of lists.  

> *Orie Steele:* +1 Manu, it would be good to comment on "how" to do the things... and how to use `@context` to do them.

**Manu Sporny:** We need a concrete use cases.  
… I'm fine to keep it open.  
… If someone wants to expose an array of ordered arrays that can do it one way and if they want RDF semantics they can do it another way.  
… I'm not volunteering to do it.  
… We need someone who knows what we're trying to accomplish.  

> *Orie Steele:* +1 to providing guidance on how to model things in JSON-LD.

**Manu Sporny:** There's lots of other things we can do that we're not providing guidance for in the spec.  

> *Orie Steele:* especially the comment about how to model arrays or arrays, and maps of maps.

**Manu Sporny:** If you want JSON semantics, use `@json`.  

**Michael Prorock:** A common example is arrays of arrays of (longitude, latitude) or (longitude, latitude, altitude).  

> *Orie Steele:* +1 to pointing to geojsonld when providing guidance on locations in JSON-LD.

> *Michael Prorock:* See [https://geojson.org/geojson-ld/](https://geojson.org/geojson-ld/).

**Manu Sporny:** There's a Geo JSON-LD spec.  

**Michael Prorock:** This would be a very helpful example.  

</details>

**Comment by msporny** [comment]
PR #1258 has been raised to address this issue. This issue will be closed when PR #1258 has been merged.
